File: Golang/Medium/3147_Taking_Maximum_Energy_From_the_Mystic_Dungeon/solution.go
package takingmaximumenergyfromthemysticdungeon

import "math"

func maximumEnergy(energy []int, k int) int {
	maxEnergy := math.MinInt
	memo := make(map[int]int)
	for i, _ := range energy {
		maxEnergy = max(maxEnergy, explore(i, k, energy, memo))
	}
	return maxEnergy
}

func explore(p int, k int, energy []int, memo map[int]int) int {
	if p >= len(energy) {
		return 0
	}

	if v, exist := memo[p]; exist {
		return v
	}

	memo[p] = energy[p] + explore(p+k, k, energy, memo)
	return memo[p]
}

File: Python/Medium/3147_Taking_Maximum_Energy_From_the_Mystic_Dungeon/solution.py
class Solution:
    def maximumEnergy(self, energy: list[int], k: int) -> int:
        maxEnergy = float('-inf')
        memo = {}
        for i in range(len(energy)):
            maxEnergy = max(maxEnergy, Solution.explore(i, k, energy, memo))
        return maxEnergy

        
    def explore(p, k, energy, memo):
        if p >= len(energy):
            return 0

        if p in memo:
            return memo[p]
        memo[p] = energy[p] + Solution.explore(p+k, k, energy, memo)
        return memo[p]
